pandas/tests/arrays/test_datetimes.py

```diff
@@ -0,0 +1,96 @@
+"""
+Tests for DatetimeArray
+"""
+import pytest
+import numpy as np
+
+from pandas.core.arrays import DatetimeArrayMixin as DatetimeArray
+
+import pandas as pd
+import pandas.util.testing as tm
+
+
+class TestDatetimeArrayConstructors(object):
+
+    def test_scalar_raises_type_error(self):
+        # GH#23493
+        with pytest.raises(TypeError):
+            DatetimeArray(2)
+
+        with pytest.raises(TypeError):
+            pd.DatetimeIndex(pd.Timestamp.now())
+
+    def test_from_sequence_requires_1dim(self):
+        arr2d = np.arange(10).view('M8[s]').astype(object).reshape(2, 5)
+        with pytest.raises(TypeError):
+            DatetimeArray(arr2d)
+
+        with pytest.raises(TypeError):
+            pd.DatetimeIndex(arr2d)
+
+        arr0d = np.array(pd.Timestamp.now())
+        with pytest.raises(TypeError):
+            DatetimeArray(arr0d)
+
+        with pytest.raises(TypeError):
+            pd.DatetimeIndex(arr0d)
+
+    def test_init_from_object_dtype(self, tz_naive_fixture):
+        # GH#23493
+        tz = tz_naive_fixture
+        if tz is not None:
+            pytest.xfail(reason="Casting DatetimeIndex to object-dtype raises "
+                                "for pd.Index and is incorrect for np.array; "
+                                "GH#24391")
+
+        # arbitrary DatetimeIndex; this should work for any DatetimeIndex
+        #  with non-None freq
+        dti = pd.date_range('2016-01-1', freq='MS', periods=9, tz=tz)
+
+        # Fails because np.array(dti, dtype=object) incorrectly returns Longs
+        result = DatetimeArray(np.array(dti, dtype=object), freq='infer')
```

Comment:
I think I mentioned above, but I don't think `DatetimeArray` constructor should handle this case? 

That is what we discussed in https://github.com/pandas-dev/pandas/issues/23212 (although, @jbrockmendel , you didn't really react there)

Reply:
Hopefully I've answered this enough times in this thread.  I see no reason _not_ to handle object dtype in the DatetimeArray constructor.  Series and Index and DateFrame all handle object-dtype and lists; I find it counter-intuitive that we would have a small subset of classes that don't.

But in the short-run, we need to handle these cases if we want to share the extant tests (without significant overhaul) (which we do)

Reply:
See https://github.com/pandas-dev/pandas/pull/23493#discussion_r230850779 for my answer to a previous related comment

Reply:
Elsewhere in this thread you've said you're fine with bikeshedding this after we have a working implementation.  Has this changed in the last few minutes?